High CPU usage

Hello,

I've noticed a high CPU usage when using Lite XL and I strongly suspect a culprit on the UI drawing level.

## Setup

 - MacBook Pro 2018, OS X 12.6
 - Lite-XL from commit 9d48441685519a9497057a2af3a3e2303f86336f

To be sure, I've tried to run Lite-XL on a bare setup:

 - Removed `~/.config/lite-xl`
 - Removed plugins from `share/lite-xl/plugins`

## Observations

| Context | CPU usage |
| ---- | ---- |
| Home screen (no document open) | 50% |
| Folder open with two documents on screen | 58% |
| Normal usage at 1 FPS | 15% |
| With base plugins reactivated and few additional plugins (gitdiff/gitstatus/linenumbers/indentguide/...) | 83% |
| App put in background after a few seconds | 0.5% |

Also, I've added a `print()` in `core.step()` to check how many times the drawing cycle is called and it doesn't seem to stop, unless the app has been in background for a while.

## Notes

 - I feel there are unecessary cycles happening on the UI drawing procedure.
 - You may have noticed that a normal usage at 1 FPS goes up to 15% which is not optimal. Other IDEs/editor (Zed/Atom/Sublime) hardly go above 5% when idle.
 - Other critical processes in the background might cause the 15% CPU usage at 1 FPS but I don't have the knowledge yet to analyse all the callstack.

Thanks for considering this issue and please forgive me if it has already been taken into account.

> are you compiling with -Drenderer=true ?

Hard to say. The compilation fails due to the amount of errors due to the SDL package from Brew (SDL 2.26).

<details><summary>See my output</summary>

```sh
[4/14] Compiling C object src/lite-xl.p/rencache.c.o
FAILED: src/lite-xl.p/rencache.c.o 
cc -Isrc/lite-xl.p -Isrc -I../src -I/usr/local/include/lua -I/usr/local/include -I/usr/local/include/SDL2 -I/usr/local/opt/freetype/include/freetype2 -I/usr/local/Cellar/pcre2/10.40/include -fcolor-diagnostics -Wall -Winvalid-pch -std=gnu11 -O0 -g -D_THREAD_SAFE -DSDL_MAIN_HANDLED -DPCRE2_STATIC -Ddirmonitor_backend=dummy -Drenderer=true -DLITE_USE_SDL_RENDERER -MD -MQ src/lite-xl.p/rencache.c.o -MF src/lite-xl.p/rencache.c.o.d -o src/lite-xl.p/rencache.c.o -c ../src/rencache.c
In file included from ../src/rencache.c:19:
In file included from ../src/rencache.h:6:
In file included from ../src/renderer.h:4:
In file included from /usr/local/include/SDL2/SDL.h:55:
/usr/local/include/SDL2/SDL_render.h:207:69: error: expected ')'
                                SDL_Window **window, SDL_Renderer **renderer);
                                                                    ^
<command line>:5:18: note: expanded from here
#define renderer true
                 ^
/Library/Developer/CommandLineTools/usr/lib/clang/14.0.0/include/stdbool.h:16:14: note: expanded from macro 'true'
#define true 1
             ^
/usr/local/include/SDL2/SDL_render.h:205:56: note: to match this '('
extern DECLSPEC int SDLCALL SDL_CreateWindowAndRenderer(
                                                       ^
/usr/local/include/SDL2/SDL_render.h:273:72: error: expected ')'
extern DECLSPEC SDL_Window * SDLCALL SDL_RenderGetWindow(SDL_Renderer *renderer);
                                                                       ^
<command line>:5:18: note: expanded from here
#define renderer true
                 ^
/Library/Developer/CommandLineTools/usr/lib/clang/14.0.0/include/stdbool.h:16:14: note: expanded from macro 'true'
#define true 1
             ^
/usr/local/include/SDL2/SDL_render.h:273:57: note: to match this '('
extern DECLSPEC SDL_Window * SDLCALL SDL_RenderGetWindow(SDL_Renderer *renderer);
                                                        ^
/usr/local/include/SDL2/SDL_render.h:288:64: error: expected ')'
extern DECLSPEC int SDLCALL SDL_GetRendererInfo(SDL_Renderer * renderer,
                                                               ^
<command line>:5:18: note: expanded from here
#define renderer true
                 ^
/Library/Developer/CommandLineTools/usr/lib/clang/14.0.0/include/stdbool.h:16:14: note: expanded from macro 'true'
#define true 1
             ^
/usr/local/include/SDL2/SDL_render.h:288:48: note: to match this '('
extern DECLSPEC int SDLCALL SDL_GetRendererInfo(SDL_Renderer * renderer,
                                               ^
/usr/local/include/SDL2/SDL_render.h:308:70: error: expected ')'
extern DECLSPEC int SDLCALL SDL_GetRendererOutputSize(SDL_Renderer * renderer,
                                                                     ^
<command line>:5:18: note: expanded from here
#define renderer true
                 ^
/Library/Developer/CommandLineTools/usr/lib/clang/14.0.0/include/stdbool.h:16:14: note: expanded from macro 'true'
#define true 1
             ^
/usr/local/include/SDL2/SDL_render.h:308:54: note: to match this '('
extern DECLSPEC int SDLCALL SDL_GetRendererOutputSize(SDL_Renderer * renderer,
                                                     ^
/usr/local/include/SDL2/SDL_render.h:333:72: error: expected ')'
extern DECLSPEC SDL_Texture * SDLCALL SDL_CreateTexture(SDL_Renderer * renderer,
                                                                       ^
<command line>:5:18: note: expanded from here
#define renderer true
                 ^
/Library/Developer/CommandLineTools/usr/lib/clang/14.0.0/include/stdbool.h:16:14: note: expanded from macro 'true'
#define true 1
             ^
/usr/local/include/SDL2/SDL_render.h:333:56: note: to match this '('
extern DECLSPEC SDL_Texture * SDLCALL SDL_CreateTexture(SDL_Renderer * renderer,
                                                       ^
/usr/local/include/SDL2/SDL_render.h:362:83: error: expected ')'
extern DECLSPEC SDL_Texture * SDLCALL SDL_CreateTextureFromSurface(SDL_Renderer * renderer, SDL_Surface * surface);
                                                                                  ^
<command line>:5:18: note: expanded from here
#define renderer true
                 ^
/Library/Developer/CommandLineTools/usr/lib/clang/14.0.0/include/stdbool.h:16:14: note: expanded from macro 'true'
#define true 1
             ^
/usr/local/include/SDL2/SDL_render.h:362:67: note: to match this '('
extern DECLSPEC SDL_Texture * SDLCALL SDL_CreateTextureFromSurface(SDL_Renderer * renderer, SDL_Surface * surface);
                                                                  ^
/usr/local/include/SDL2/SDL_render.h:755:74: error: expected ')'
extern DECLSPEC SDL_bool SDLCALL SDL_RenderTargetSupported(SDL_Renderer *renderer);
                                                                         ^
<command line>:5:18: note: expanded from here
#define renderer true
                 ^
/Library/Developer/CommandLineTools/usr/lib/clang/14.0.0/include/stdbool.h:16:14: note: expanded from macro 'true'
#define true 1
             ^
/usr/local/include/SDL2/SDL_render.h:755:59: note: to match this '('
extern DECLSPEC SDL_bool SDLCALL SDL_RenderTargetSupported(SDL_Renderer *renderer);
                                                          ^
/usr/local/include/SDL2/SDL_render.h:779:63: error: expected ')'
extern DECLSPEC int SDLCALL SDL_SetRenderTarget(SDL_Renderer *renderer,
                                                              ^
<command line>:5:18: note: expanded from here
#define renderer true
                 ^
/Library/Developer/CommandLineTools/usr/lib/clang/14.0.0/include/stdbool.h:16:14: note: expanded from macro 'true'
#define true 1
             ^
/usr/local/include/SDL2/SDL_render.h:779:48: note: to match this '('
extern DECLSPEC int SDLCALL SDL_SetRenderTarget(SDL_Renderer *renderer,
                                               ^
/usr/local/include/SDL2/SDL_render.h:795:73: error: expected ')'
extern DECLSPEC SDL_Texture * SDLCALL SDL_GetRenderTarget(SDL_Renderer *renderer);
                                                                        ^
<command line>:5:18: note: expanded from here
#define renderer true
                 ^
/Library/Developer/CommandLineTools/usr/lib/clang/14.0.0/include/stdbool.h:16:14: note: expanded from macro 'true'
#define true 1
             ^
/usr/local/include/SDL2/SDL_render.h:795:58: note: to match this '('
extern DECLSPEC SDL_Texture * SDLCALL SDL_GetRenderTarget(SDL_Renderer *renderer);
                                                         ^
/usr/local/include/SDL2/SDL_render.h:823:69: error: expected ')'
extern DECLSPEC int SDLCALL SDL_RenderSetLogicalSize(SDL_Renderer * renderer, int w, int h);
                                                                    ^
<command line>:5:18: note: expanded from here
#define renderer true
                 ^
/Library/Developer/CommandLineTools/usr/lib/clang/14.0.0/include/stdbool.h:16:14: note: expanded from macro 'true'
#define true 1
             ^
/usr/local/include/SDL2/SDL_render.h:823:53: note: to match this '('
extern DECLSPEC int SDLCALL SDL_RenderSetLogicalSize(SDL_Renderer * renderer, int w, int h);
                                                    ^
/usr/local/include/SDL2/SDL_render.h:844:70: error: expected ')'
extern DECLSPEC void SDLCALL SDL_RenderGetLogicalSize(SDL_Renderer * renderer, int *w, int *h);
                                                                     ^
<command line>:5:18: note: expanded from here
#define renderer true
                 ^
/Library/Developer/CommandLineTools/usr/lib/clang/14.0.0/include/stdbool.h:16:14: note: expanded from macro 'true'
#define true 1
             ^
/usr/local/include/SDL2/SDL_render.h:844:54: note: to match this '('
extern DECLSPEC void SDLCALL SDL_RenderGetLogicalSize(SDL_Renderer * renderer, int *w, int *h);
                                                     ^
/usr/local/include/SDL2/SDL_render.h:863:70: error: expected ')'
extern DECLSPEC int SDLCALL SDL_RenderSetIntegerScale(SDL_Renderer * renderer,
                                                                     ^
<command line>:5:18: note: expanded from here
#define renderer true
                 ^
/Library/Developer/CommandLineTools/usr/lib/clang/14.0.0/include/stdbool.h:16:14: note: expanded from macro 'true'
#define true 1
             ^
/usr/local/include/SDL2/SDL_render.h:863:54: note: to match this '('
extern DECLSPEC int SDLCALL SDL_RenderSetIntegerScale(SDL_Renderer * renderer,
                                                     ^
/usr/local/include/SDL2/SDL_render.h:877:75: error: expected ')'
extern DECLSPEC SDL_bool SDLCALL SDL_RenderGetIntegerScale(SDL_Renderer * renderer);
                                                                          ^
<command line>:5:18: note: expanded from here
#define renderer true
                 ^
/Library/Developer/CommandLineTools/usr/lib/clang/14.0.0/include/stdbool.h:16:14: note: expanded from macro 'true'
#define true 1
             ^
/usr/local/include/SDL2/SDL_render.h:877:59: note: to match this '('
extern DECLSPEC SDL_bool SDLCALL SDL_RenderGetIntegerScale(SDL_Renderer * renderer);
                                                          ^
/usr/local/include/SDL2/SDL_render.h:895:66: error: expected ')'
extern DECLSPEC int SDLCALL SDL_RenderSetViewport(SDL_Renderer * renderer,
                                                                 ^
<command line>:5:18: note: expanded from here
#define renderer true
                 ^
/Library/Developer/CommandLineTools/usr/lib/clang/14.0.0/include/stdbool.h:16:14: note: expanded from macro 'true'
#define true 1
             ^
/usr/local/include/SDL2/SDL_render.h:895:50: note: to match this '('
extern DECLSPEC int SDLCALL SDL_RenderSetViewport(SDL_Renderer * renderer,
                                                 ^
/usr/local/include/SDL2/SDL_render.h:908:67: error: expected ')'
extern DECLSPEC void SDLCALL SDL_RenderGetViewport(SDL_Renderer * renderer,
                                                                  ^
<command line>:5:18: note: expanded from here
#define renderer true
                 ^
/Library/Developer/CommandLineTools/usr/lib/clang/14.0.0/include/stdbool.h:16:14: note: expanded from macro 'true'
#define true 1
             ^
/usr/local/include/SDL2/SDL_render.h:908:51: note: to match this '('
extern DECLSPEC void SDLCALL SDL_RenderGetViewport(SDL_Renderer * renderer,
                                                  ^
/usr/local/include/SDL2/SDL_render.h:926:66: error: expected ')'
extern DECLSPEC int SDLCALL SDL_RenderSetClipRect(SDL_Renderer * renderer,
                                                                 ^
<command line>:5:18: note: expanded from here
#define renderer true
                 ^
/Library/Developer/CommandLineTools/usr/lib/clang/14.0.0/include/stdbool.h:16:14: note: expanded from macro 'true'
#define true 1
             ^
/usr/local/include/SDL2/SDL_render.h:926:50: note: to match this '('
extern DECLSPEC int SDLCALL SDL_RenderSetClipRect(SDL_Renderer * renderer,
                                                 ^
/usr/local/include/SDL2/SDL_render.h:942:67: error: expected ')'
extern DECLSPEC void SDLCALL SDL_RenderGetClipRect(SDL_Renderer * renderer,
                                                                  ^
<command line>:5:18: note: expanded from here
#define renderer true
                 ^
/Library/Developer/CommandLineTools/usr/lib/clang/14.0.0/include/stdbool.h:16:14: note: expanded from macro 'true'
#define true 1
             ^
/usr/local/include/SDL2/SDL_render.h:942:51: note: to match this '('
extern DECLSPEC void SDLCALL SDL_RenderGetClipRect(SDL_Renderer * renderer,
                                                  ^
/usr/local/include/SDL2/SDL_render.h:957:73: error: expected ')'
extern DECLSPEC SDL_bool SDLCALL SDL_RenderIsClipEnabled(SDL_Renderer * renderer);
                                                                        ^
<command line>:5:18: note: expanded from here
#define renderer true
                 ^
/Library/Developer/CommandLineTools/usr/lib/clang/14.0.0/include/stdbool.h:16:14: note: expanded from macro 'true'
#define true 1
             ^
/usr/local/include/SDL2/SDL_render.h:957:57: note: to match this '('
extern DECLSPEC SDL_bool SDLCALL SDL_RenderIsClipEnabled(SDL_Renderer * renderer);
                                                        ^
/usr/local/include/SDL2/SDL_render.h:982:63: error: expected ')'
extern DECLSPEC int SDLCALL SDL_RenderSetScale(SDL_Renderer * renderer,
                                                              ^
<command line>:5:18: note: expanded from here
#define renderer true
                 ^
/Library/Developer/CommandLineTools/usr/lib/clang/14.0.0/include/stdbool.h:16:14: note: expanded from macro 'true'
#define true 1
             ^
/usr/local/include/SDL2/SDL_render.h:982:47: note: to match this '('
extern DECLSPEC int SDLCALL SDL_RenderSetScale(SDL_Renderer * renderer,
                                              ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.

```

</details>

Also, I'm using the `build-packages.sh` script to compile.

Am I missing something?

--- a/src/api/dirmonitor/fsevents.c
+++ b/src/api/dirmonitor/fsevents.c
@@ -108,6 +108,9 @@ int get_changes_dirmonitor(
   results = monitor->count;
   SDL_UnlockMutex(monitor->lock);
 
+  if (monitor->count <= 0)
+    SDL_Delay(100);
+
   return results;
 }
 

--- a/src/api/dirmonitor/kqueue.c
+++ b/src/api/dirmonitor/kqueue.c
@@ -24,7 +24,7 @@ void deinit_dirmonitor(struct dirmonitor_internal* monitor) {
 
 
 int get_changes_dirmonitor(struct dirmonitor_internal* monitor, char* buffer, int buffer_size) {
-  struct timespec ts = { 0, 10 * 1000000 }; // 10 ms
+  struct timespec ts = { 0, 100 * 1000000 }; // 100 ms
 
   int nev = kevent(monitor->fd, NULL, 0, (struct kevent*)buffer, buffer_size / sizeof(kevent), &ts);
   if (nev == -1)
@@ -45,7 +45,7 @@ int translate_changes_dirmonitor(struct dirmonitor_internal* monitor, char* buff
 int add_dirmonitor(struct dirmonitor_internal* monitor, const char* path) {
   int fd = open(path, O_RDONLY);
   struct kevent change;
-  
+
   // a timeout of zero should make kevent return immediately
   struct timespec ts = { 0, 0 }; // 0 s
 

